File: elixir/palindrome-products/palindrome_products.exs
defmodule Palindromes do
  @doc """
  Generates all palindrome products from an optionally given min factor (or 1) to a given max factor.
  """
  @spec generate(non_neg_integer, non_neg_integer) :: map
  def generate(max_factor, min_factor \\ 1)

  def generate(max_factor, min_factor) do
    generate(min_factor, min_factor, max_factor, %{})
  end

  def generate(x, y, max, acc) do
    acc = cond do
      is_palindrome(x*y) -> Map.update(acc, x*y, [[x,y]], &([[x,y]|&1]))
      true               -> acc
    end
    cond do
      x >= max && y >= max   -> acc
      y <  max               -> generate(x, y+1, max, acc)
      y >= max               -> generate(x+1, x+1, max, acc)
    end
  end

  defp is_palindrome(number) do
    number_string = Integer.to_string(number)
    number_string == String.reverse(number_string)
  end
end

File: javascript/palindrome-products/palindrome-products.js
var Palindromes = function(params) {
  this.maxFactor = params.maxFactor;
  this.minFactor = params.minFactor || 1;
  this.results = {};
};

Palindromes.prototype.generate = function() {
  for (var x=this.minFactor; x<=this.maxFactor; x++) {
    for (var y=x; y<=this.maxFactor; y++) {
      if (this.isPalindrome(x*y)) {
        if (!this.results[x*y]) {
          this.results[x*y] = { value: x*y, factors: [[x,y]] };
        } else {
          this.results[x*y].factors.push([x,y]);
        }
      }
    }
  }
};

Palindromes.prototype.largest = function() {
  let out = {value: 0};
  Object.keys(this.results).forEach((k) => {
    if (this.results[k].value > out.value) {
      out = this.results[k];
    }
  });
  return out;
};

Palindromes.prototype.smallest = function() {
  let out = {value: Number.MAX_VALUE};
  Object.keys(this.results).forEach((k) => {
    if (this.results[k].value < out.value) {
      out = this.results[k];
    }
  });
  return out;
};

Palindromes.prototype.isPalindrome = function(input) {
  input = input.toString();
  return input == input.split('').reverse().join('');
};

module.exports = Palindromes;
